sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/MultimapUserState.java

```diff
@@ -0,0 +1,287 @@
+ */
+public class MultimapUserState<K, V> {
+
+  private final BeamFnStateClient beamFnStateClient;
+  private final Coder<K> mapKeyCoder;
+  private final Coder<V> valueCoder;
+  private final String stateId;
+  private final StateRequest keysStateRequest;
+  private final StateRequest userStateRequest;
+
+  private boolean isClosed;
+  private boolean isCleared;
+  // Pending updates to persistent storage
+  private Set<K> pendingRemoves = Sets.newHashSet();
+  private Multimap<K, V> pendingAdds = ArrayListMultimap.create();
+  // Map keys with no values in persistent storage
+  private Set<K> negativeCache = Sets.newHashSet();
+  // Values retrieved from persistent storage
+  private Multimap<K, V> persistedValues = ArrayListMultimap.create();
+  private @Nullable Iterable<K> persistedKeys = null;
+
+  public MultimapUserState(
+      BeamFnStateClient beamFnStateClient,
+      String instructionId,
+      String pTransformId,
+      String stateId,
+      ByteString encodedWindow,
+      ByteString encodedKey,
+      Coder<K> mapKeyCoder,
+      Coder<V> valueCoder) {
+    this.beamFnStateClient = beamFnStateClient;
+    this.mapKeyCoder = mapKeyCoder;
+    this.valueCoder = valueCoder;
+    this.stateId = stateId;
+
+    StateRequest.Builder keysStateRequestBuilder = StateRequest.newBuilder();
+    keysStateRequestBuilder
+        .setInstructionId(instructionId)
+        .getStateKeyBuilder()
+        .getMultimapKeysUserStateBuilder()
+        .setTransformId(pTransformId)
+        .setUserStateId(stateId)
+        .setKey(encodedKey)
+        .setWindow(encodedWindow);
+    keysStateRequest = keysStateRequestBuilder.build();
+
+    StateRequest.Builder userStateRequestBuilder = StateRequest.newBuilder();
+    userStateRequestBuilder
+        .setInstructionId(instructionId)
+        .getStateKeyBuilder()
+        .getMultimapUserStateBuilder()
+        .setTransformId(pTransformId)
+        .setUserStateId(stateId)
+        .setWindow(encodedWindow)
+        .setKey(encodedKey);
+    userStateRequest = userStateRequestBuilder.build();
+  }
+
+  public void clear() {
+    checkState(
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+
+    isCleared = true;
+    persistedValues = ArrayListMultimap.create();
+    persistedKeys = null;
+    pendingRemoves = Sets.newHashSet();
+    pendingAdds = ArrayListMultimap.create();
+  }
+
+  /*
+   * Returns an iterable of the values associated with key in this multimap, if any.
+   * If there are no values, this returns an empty collection, not null.
+   */
+  public Iterable<V> get(@NonNull K key) {
+    checkState(
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+
+    Collection<V> pendingValues =
+        Collections.unmodifiableCollection(Lists.newArrayList(pendingAdds.get(key)));
```

Comment:
For a follow-up change:
Note that if we use a `Map<K, List<V>>` instead of a `ArrayListMultimap` we can control the lifecycle of the List and how it is cleared/replaced allowing us to replace the copy being done via `Lists.newArrayList` to rely on `List#sublist` to restrict the range. Clearing the `Map` will remove the references to the lists and any current _sub lists_ that the user is holding onto will remain alive. `ArrayListMultimap` doesn't explicitly state these guarantees. 

It is harder to do something like this for _keys_ but also possible.